README.md
# day_night_theme_flutter
A Flutter plugin that helps you to automatically change the theme of the app with sunrise and sunset. Just specify the light and dark theme to use, and you are all set. You can use your custom sunrise and sunset time too.

<img src = "https://raw.githubusercontent.com/maheshbosctechlabs/day_night_theme_flutter/main/plugin_preview.png" height= "700"/>

## How to use?

1. Add the latest version of the package in your ```pubspec.yaml```
2. Wrap the ```MaterialApp``` with ```DayNightTheme``` widget.

```
void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // Light theme of your app
  final ThemeData _lightTheme = ThemeData(
    primarySwatch: Colors.blue,
    visualDensity: VisualDensity.adaptivePlatformDensity,
    brightness: Brightness.light,
  );
  // Dark theme of your app
  final ThemeData _darkTheme = ThemeData(
    primarySwatch: Colors.brown,
    visualDensity: VisualDensity.adaptivePlatformDensity,
    brightness: Brightness.dark,
  );

  @override
  Widget build(BuildContext context) {
    return DayNightTheme(
      // specify your themes
      darkTheme: _darkTheme,
      lightTheme: _lightTheme,
      // sunrise time in 24 hours format
      sunriseHour: 6,
      sunriseMinutes: 30,
      // sunset time in 24 hours format
      sunsetHour: 19,
      sunsetMinutes: 0,
      builder: (selectedTheme) {
        return MaterialApp(
          title: 'Flutter Demo',
          // apply the theme
          theme: selectedTheme,
          home: MyHomePage(title: 'Flutter Demo Home Page'),
        );
      },
    );
  }
}
```

## Supporters
[<img src="https://bosctechlabs.com/wp-content/themes/bosc/assets/images/logo.png" />](https://bosctechlabs.com/)


CHANGELOG.md
## [0.0.5]

* Null safety added

## [0.0.4]

* minor changes

## [0.0.3]

* readme updated

## [0.0.2]

* pub.dev analysis related changes
* example added

## [0.0.1]

* Initial Version of the library.


test/day_night_theme_flutter_test.dart
import 'package:day_night_theme_flutter/extensions/date_time_extension.dart';
import 'package:day_night_theme_flutter/model/theme_changer.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  final ThemeData _lightTheme = ThemeData(
    primarySwatch: Colors.blue,
    visualDensity: VisualDensity.adaptivePlatformDensity,
    brightness: Brightness.light,
  );

  final ThemeData _darkTheme = ThemeData(
    primarySwatch: Colors.brown,
    visualDensity: VisualDensity.adaptivePlatformDensity,
    brightness: Brightness.dark,
  );
  test("isLightTheme false before sunrise", () {
    DateTime now = DateTime.now();
    CustomDateTime.customTime =
        DateTime(now.year, now.month, now.day, 2, 11, 55);

    DayNightThemeChanger themeChanger =
        DayNightThemeChanger(_lightTheme, _darkTheme, 2, 13, 2, 12);

    expect(themeChanger.isLightTheme, false);
  });

  test("isLightTheme true during the day", () {
    DateTime now = DateTime.now();
    CustomDateTime.customTime =
        DateTime(now.year, now.month, now.day, 2, 12, 30);

    DayNightThemeChanger themeChanger =
        DayNightThemeChanger(_lightTheme, _darkTheme, 2, 13, 2, 12);

    expect(themeChanger.isLightTheme, true);
  });

  test("isLightTheme false after sunset", () {
    DateTime now = DateTime.now();
    CustomDateTime.customTime =
        DateTime(now.year, now.month, now.day, 2, 14, 30);

    DayNightThemeChanger themeChanger =
        DayNightThemeChanger(_lightTheme, _darkTheme, 2, 13, 2, 12);

    expect(themeChanger.isLightTheme, false);
  });

  test("theme changes from dark to light", () async {
    DateTime now = DateTime.now();
    CustomDateTime.customTime =
        DateTime(now.year, now.month, now.day, 2, 11, 55);

    DayNightThemeChanger themeChanger =
        DayNightThemeChanger(_lightTheme, _darkTheme, 2, 13, 2, 12);

    expect(themeChanger.isLightTheme, false);

    CustomDateTime.customTime =
        DateTime(now.year, now.month, now.day, 2, 12, 30);

    await Future.delayed(Duration(seconds: 10));

    expect(themeChanger.isLightTheme, true);
  });

  test("theme changes from light to dark", () async {
    DateTime now = DateTime.now();

    CustomDateTime.customTime =
        DateTime(now.year, now.month, now.day, 2, 12, 55);

    DayNightThemeChanger themeChanger =
        DayNightThemeChanger(_lightTheme, _darkTheme, 2, 13, 2, 12);

    expect(themeChanger.isLightTheme, true);

    CustomDateTime.customTime =
        DateTime(now.year, now.month, now.day, 2, 13, 30);

    await Future.delayed(Duration(seconds: 10));

    expect(themeChanger.isLightTheme, false);
  });

  test('sunrise time is less than sunset time', () {
    expect(() => DayNightThemeChanger(_lightTheme, _darkTheme, 1, 13, 2, 14),
        throwsA(isAssertionError));
  });

  test(
      'sunrise hour is same as sunset time but sunrise minutes should be less than sunset minutes',
      () {
    expect(() => DayNightThemeChanger(_lightTheme, _darkTheme, 2, 13, 2, 14),
        throwsA(isAssertionError));
  });

  test('sunrise time is not as same as sunset time', () {
    expect(() => DayNightThemeChanger(_lightTheme, _darkTheme, 2, 14, 2, 14),
        throwsA(isAssertionError));
  });
}


test/day_night_theme_flutter_widget_test.dart
import 'package:day_night_theme_flutter/day_night_theme_flutter.dart';
import 'package:day_night_theme_flutter/extensions/date_time_extension.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  final ThemeData _lightTheme = ThemeData(
    primarySwatch: Colors.blue,
    visualDensity: VisualDensity.adaptivePlatformDensity,
    brightness: Brightness.light,
  );

  final ThemeData _darkTheme = ThemeData(
    primarySwatch: Colors.brown,
    visualDensity: VisualDensity.adaptivePlatformDensity,
    brightness: Brightness.dark,
  );

  final scaffoldKey = ValueKey('Scaffold');

  testWidgets(
    "Theme is dark before sunrise",
    (WidgetTester tester) async {
      DateTime now = DateTime.now();
      CustomDateTime.customTime = DateTime(now.year, now.month, now.day, 2, 11);
      Widget testingWidget = DayNightTheme(
        builder: (selectedTheme) {
          return MaterialApp(
            title: 'Flutter Demo',
            theme: selectedTheme,
            home: Scaffold(
              key: scaffoldKey,
              appBar: AppBar(
                title: Text('Widget Testing'),
              ),
            ),
          );
        },
        darkTheme: _darkTheme,
        lightTheme: _lightTheme,
        sunriseHour: 2,
        sunriseMinutes: 12,
        sunsetHour: 2,
        sunsetMinutes: 13,
      );
      await tester.pumpWidget(testingWidget);
      expect(Theme.of(tester.element(find.byKey(scaffoldKey))).brightness,
          _darkTheme.brightness);
    },
  );

  testWidgets(
    "Theme is light during the day",
    (WidgetTester tester) async {
      DateTime now = DateTime.now();
      CustomDateTime.customTime =
          DateTime(now.year, now.month, now.day, 2, 12, 30);
      Widget testingWidget = DayNightTheme(
        builder: (selectedTheme) {
          return MaterialApp(
            title: 'Flutter Demo',
            theme: selectedTheme,
            home: Scaffold(
              key: scaffoldKey,
              appBar: AppBar(
                title: Text('Widget Testing'),
              ),
            ),
          );
        },
        darkTheme: _darkTheme,
        lightTheme: _lightTheme,
        sunriseHour: 2,
        sunriseMinutes: 12,
        sunsetHour: 2,
        sunsetMinutes: 13,
      );
      await tester.pumpWidget(testingWidget);

      expect(Theme.of(tester.element(find.byKey(scaffoldKey))).brightness,
          _lightTheme.brightness);
    },
  );

  testWidgets(
    "Theme is dark after sunset",
    (WidgetTester tester) async {
      DateTime now = DateTime.now();
      CustomDateTime.customTime = DateTime(now.year, now.month, now.day, 2, 14);
      Widget testingWidget = DayNightTheme(
        builder: (selectedTheme) {
          return MaterialApp(
            title: 'Flutter Demo',
            theme: selectedTheme,
            home: Scaffold(
              key: scaffoldKey,
              appBar: AppBar(
                title: Text('Widget Testing'),
              ),
            ),
          );
        },
        darkTheme: _darkTheme,
        lightTheme: _lightTheme,
        sunriseHour: 2,
        sunriseMinutes: 12,
        sunsetHour: 2,
        sunsetMinutes: 13,
      );
      await tester.pumpWidget(testingWidget);

      expect(Theme.of(tester.element(find.byKey(scaffoldKey))).brightness,
          _darkTheme.brightness);
    },
  );

  testWidgets(
    "sunrise time is less than sunset time",
    (WidgetTester tester) async {
      DateTime now = DateTime.now();
      CustomDateTime.customTime = DateTime(now.year, now.month, now.day, 2, 14);

      final Matcher throwsAssertionError = throwsA(isAssertionError);

      expect(
        () => DayNightTheme(
          builder: (selectedTheme) {
            return Container();
          },
          darkTheme: _darkTheme,
          lightTheme: _lightTheme,
          sunriseHour: 4,
          sunriseMinutes: 12,
          sunsetHour: 2,
          sunsetMinutes: 13,
        ),
        throwsAssertionError,
      );
    },
  );

  testWidgets(
    "sunrise hour is same as sunset time but sunrise minutes should be less than sunset minutes",
    (WidgetTester tester) async {
      DateTime now = DateTime.now();
      CustomDateTime.customTime = DateTime(now.year, now.month, now.day, 2, 14);

      final Matcher throwsAssertionError = throwsA(isAssertionError);

      expect(
        () => DayNightTheme(
          builder: (selectedTheme) {
            return Container();
          },
          darkTheme: _darkTheme,
          lightTheme: _lightTheme,
          sunriseHour: 2,
          sunriseMinutes: 14,
          sunsetHour: 2,
          sunsetMinutes: 13,
        ),
        throwsAssertionError,
      );
    },
  );

  testWidgets(
    "sunrise time is not as same as sunset time",
    (WidgetTester tester) async {
      DateTime now = DateTime.now();
      CustomDateTime.customTime = DateTime(now.year, now.month, now.day, 2, 14);

      final Matcher throwsAssertionError = throwsA(isAssertionError);

      expect(
        () => DayNightTheme(
          builder: (selectedTheme) {
            return Container();
          },
          darkTheme: _darkTheme,
          lightTheme: _lightTheme,
          sunriseHour: 2,
          sunriseMinutes: 12,
          sunsetHour: 2,
          sunsetMinutes: 12,
        ),
        throwsAssertionError,
      );
    },
  );
}


lib/day_night_theme.dart
import 'package:day_night_theme_flutter/extensions/date_time_extension.dart';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

import 'package:day_night_theme_flutter/model/theme_changer.dart';

class DayNightTheme extends StatefulWidget {
  final Widget Function(ThemeData currentTheme) builder;

  final ThemeData lightTheme;

  final ThemeData darkTheme;

  /// Sunrise hour in 24 hours format
  /// default is 6 in the morning
  final int sunriseHour;

  /// Sunrise minutes
  /// default is 0
  final int sunriseMinutes;

  /// Sunset hour in 24 hours format
  /// default is 19
  final int sunsetHour;

  /// Sunset minutes
  /// default is 0
  final int sunsetMinutes;

  DayNightTheme({
    required this.builder,
    required this.darkTheme,
    required this.lightTheme,
    this.sunriseHour = 6,
    this.sunriseMinutes = 0,
    this.sunsetMinutes = 0,
    this.sunsetHour = 19,
  }) {
    DateTime now = CustomDateTime.current;

    DateTime sunriseTime =
        DateTime(now.year, now.month, now.day, sunriseHour, sunriseMinutes);
    DateTime sunsetTime =
        DateTime(now.year, now.month, now.day, sunsetHour, sunsetMinutes);

    assert(sunriseTime.isBefore(sunsetTime),
        'sunrise time must be less than sunset time');
  }

  @override
  _DayNightThemeState createState() => _DayNightThemeState();
}

class _DayNightThemeState extends State<DayNightTheme> {
  late DayNightThemeChanger _themeChanger;

  @override
  void initState() {
    super.initState();
    _themeChanger = DayNightThemeChanger(
      widget.lightTheme,
      widget.darkTheme,
      widget.sunsetHour,
      widget.sunsetMinutes,
      widget.sunriseHour,
      widget.sunriseMinutes,
    );
  }

  @override
  void dispose() {
    _themeChanger.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return ChangeNotifierProvider<DayNightThemeChanger>.value(
      value: _themeChanger,
      builder: (context, child) {
        final themeChanger = Provider.of<DayNightThemeChanger>(context);

        return widget.builder(themeChanger.selectedTheme);
      },
    );
  }
}


lib/model/theme_changer.dart
import 'dart:async';

import 'package:day_night_theme_flutter/extensions/date_time_extension.dart';
import 'package:flutter/material.dart';

class DayNightThemeChanger extends ChangeNotifier {
  late ThemeData _selectedTheme;

  final ThemeData lightTheme;

  final ThemeData darkTheme;

  /// Morning hour in 24 hours format
  /// default is 6 in the morning
  final int sunriseHour;

  /// Morning minutes
  /// default in 0
  final int sunriseMinutes;

  /// Evening start hour in 24 hours format
  /// default is 19
  final int sunsetHour;

  /// Evening start minutes
  /// default in 0
  final int sunsetMinutes;

  /// Timer will be required to change the theme after some time.
  Timer? _timer;

  DayNightThemeChanger(
    this.lightTheme,
    this.darkTheme,
    this.sunsetHour,
    this.sunsetMinutes,
    this.sunriseHour,
    this.sunriseMinutes,
  ) {
    DateTime now = CustomDateTime.current;
    DateTime sunriseTime =
        DateTime(now.year, now.month, now.day, sunriseHour, sunriseMinutes);
    DateTime sunsetTime =
        DateTime(now.year, now.month, now.day, sunsetHour, sunsetMinutes);
    assert(sunriseTime.isBefore(sunsetTime),
        'sunrise time must be less than sunset time');
    _init();
  }

  _init() {
    DateTime now = CustomDateTime.current;

    // Dark theme start time evening 7 pm
    DateTime sunsetTime = DateTime(
      now.year,
      now.month,
      now.day,
      this.sunsetHour,
      this.sunsetMinutes,
    );
    // Dark theme to Light switch happen at 6 am
    DateTime sunriseTime = DateTime(
      now.year,
      now.month,
      now.day,
      this.sunriseHour,
      this.sunriseMinutes,
    );

    // time left to change the theme
    int timerSeconds;
    if (now.isAfter(sunriseTime) && now.isBefore(sunsetTime)) {
      debugPrint('Changing theme to lightTheme');
      changeTheme(lightTheme);
      timerSeconds = sunsetTime.difference(now).inMilliseconds;
    } else if (now.isAfter(sunsetTime)) {
      debugPrint('Changing theme to darkTheme');
      changeTheme(darkTheme);
      DateTime nextDayMorningTime = sunriseTime.add(Duration(days: 1));
      timerSeconds = nextDayMorningTime.difference(now).inMilliseconds;
    } else {
      debugPrint('Changing theme to darkTheme');
      changeTheme(darkTheme);
      timerSeconds = sunriseTime.difference(now).inMilliseconds;
    }
    debugPrint('Time left to change the theme $timerSeconds');
    _timer?.cancel();
    _timer = Timer(
      Duration(milliseconds: timerSeconds),
      () {
        _init();
      },
    );
  }

  /// Manually change the theme of the app
  /// If you don't wish to change the theme at sunset / sunrise time
  /// set the ``turnOffDynamicThemeChange`` to true
  changeTheme(ThemeData newTheme, {bool turnOffDynamicThemeChange = false}) {
    _selectedTheme = newTheme;
    notifyListeners();
    if (turnOffDynamicThemeChange) {
      _timer?.cancel();
    }
  }

  /// Currently selected theme
  ThemeData get selectedTheme => _selectedTheme;

  /// returns true if current selected theme is light theme
  bool get isLightTheme => _selectedTheme == lightTheme;

  @override
  void dispose() {
    _timer?.cancel();
    super.dispose();
  }
}


example/lib/main.dart
import 'package:day_night_theme_flutter/day_night_theme_flutter.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // Light theme of your app
  final ThemeData _lightTheme = ThemeData(
    primarySwatch: Colors.blue,
    visualDensity: VisualDensity.adaptivePlatformDensity,
    brightness: Brightness.light,
  );
  // Dark theme of your app
  final ThemeData _darkTheme = ThemeData(
    primarySwatch: Colors.brown,
    visualDensity: VisualDensity.adaptivePlatformDensity,
    brightness: Brightness.dark,
  );
  @override
  Widget build(BuildContext context) {
    return DayNightTheme(
      // specify your themes
      darkTheme: _darkTheme,
      lightTheme: _lightTheme,
      // sunrise time in 24 hours format
      sunriseHour: 13,
      sunriseMinutes: 22,
      // sunset time in 24 hours format
      sunsetHour: 13,
      sunsetMinutes: 23,
      builder: (selectedTheme) {
        return MaterialApp(
          title: 'Flutter Demo',
          // apply the theme
          theme: selectedTheme,
          home: MyHomePage(title: 'Flutter Demo Home Page'),
        );
      },
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}
